statsmodels/imputation/ros.py

```diff
@@ -0,0 +1,558 @@
+        transform_in(df[result][uncensored_mask])
+    )
+
+    # pull out the slope and intercept for use later
+    slope, intercept = fit_params[:2]
+
+    # model the data based on the best-fit curve
+    df = (
+        df.assign(estimated=transform_out(slope * df['Zprelim'][censored_mask] + intercept))
+          .assign(final=lambda df: numpy.where(df[censorship], df['estimated'], df[result]))
+    )
+
+    return df
+
+
+def _do_ros(df, result, censorship, transform_in, transform_out):
+    """
+    Prepares a dataframe for, and then esimates the values of a censored
+    dataset using Regression on Order Statistics
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+
+    result : str
+        Name of the column in the dataframe that contains observed
+        values. Censored values should be set to the detection (upper)
+        limit.
+
+    censorship : str
+        Name of the column in the dataframe that indicates that a
+        result is left-censored. (i.e., True -> censored,
+        False -> uncensored)
+
+    transform_in, transform_out : callable
+        Transformations to be applied to the data prior to fitting
+        the line and after estimated values from that line. Typically,
+        `numpy.log` and `numpy.exp` are used, respectively.
+
+    Returns
+    -------
+    estimated : pandas.DataFrame
+        A new dataframe with two new columns: "estimated" and "final".
+        The "estimated" column contains of the values inferred from the
+        best-fit line. The "final" column contains the estimated values
+        only where the original results were censored, and the original
+        results everwhere else.
+
+    """
+
+    # compute the Cohn numbers
+    cohn = cohn_numbers(df, result=result, censorship=censorship)
+
+    modeled = (
+        df.pipe(_ros_sort, result=result, censorship=censorship)
+          .assign(det_limit_index=lambda df: df[result].apply(_detection_limit_index, args=(cohn,)))
+          .assign(rank=lambda df: _ros_group_rank(df, 'det_limit_index', censorship))
+          .assign(plot_pos=lambda df: plotting_positions(df, censorship, cohn))
+          .assign(Zprelim=lambda df: stats.norm.ppf(df['plot_pos']))
+          .pipe(_ros_estimate, result, censorship, transform_in, transform_out)
+    )
+
+    return modeled
+
+
+def ROS(result, censorship, df=None, min_uncensored=2,
+        max_fraction_censored=0.8, substitution_fraction=0.5,
+        transform_in=numpy.log, transform_out=numpy.exp,
+        as_array=True):
+    """
+    Impute censored dataset using Regression on Order Statistics (ROS)
+    or simple substitution if insufficient uncensored data exists.
+
+    Method described in *Nondetects and Data Analysis* by Dennis R.
+    Helsel (John Wiley, 2005) to estimate the left-censored (non-detect)
+    values of a dataset.
+
+    Parameters
+    ----------
+    result : str or array-like
+        Label of the column or the float array of censored results
```

Comment:
`results` here is a bit confusing/ambiguous
Are these other results that are used as inputs here, or does this refer to the results of this function?